File: 算法/offer66/树/重建二叉树/Solution.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
import java.util.Arrays;
public class Solution {
    public TreeNode reConstructBinaryTree(int [] pre,int [] in) {
        if (pre.length == 0 || in.length == 0)
            return null;
        TreeNode T = new TreeNode(pre[0]);
        int count = 0;

        for (int i=0;i<in.length;i++) 
        {
            if (in[i] == pre[0]) {
                count = i;
                System.out.println(count);
                break;
            }
        };
        T.left = reConstructBinaryTree(Arrays.copyOfRange(pre, 1, count + 1), Arrays.copyOfRange(in, 0, count));
        T.right = reConstructBinaryTree(Arrays.copyOfRange(pre, count+1,pre.length), Arrays.copyOfRange(in, count+1, in.length));
        return  T;
    }
}

File: 算法/offer66/树/重建二叉树/Solution.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
# 1、构建root节点
# 2、左子树的前序+中序作为输入，返回左子树根节点；同理右子树节点可以获得
# 3、根据2 递归赋值root.left和root.right；
class Solution:
    # 返回构造的TreeNode根节点
    def reConstructBinaryTree(self, pre, tin):
        # write code here
        if(len(pre) == 0): return None 
        root = TreeNode(pre[0])
        ##根据值找索引 ：ind = list.index(val)
        ind = tin.index(pre[0])
        root.left = self.reConstructBinaryTree(pre[1:ind+1],tin[0:ind])
        root.right = self.reConstructBinaryTree(pre[ind+1:],tin[ind+1:])
        return root
